chasm/tree_test.go

```diff
@@ -98,6 +98,149 @@ func (s *nodeSuite) TestNewTree() {
+			LastUpdateVersionedTransition: &persistencespb.VersionedTransition{TransitionCount: 1},
+		},
+		"child": {
+			InitialVersionedTransition:    &persistencespb.VersionedTransition{TransitionCount: 2},
+			LastUpdateVersionedTransition: &persistencespb.VersionedTransition{TransitionCount: 2},
+		},
+		"child/grandchild": {
+			InitialVersionedTransition:    &persistencespb.VersionedTransition{TransitionCount: 3},
+			LastUpdateVersionedTransition: &persistencespb.VersionedTransition{TransitionCount: 3},
+		},
+		"child/grandchild/grandgrandchild": {
+			InitialVersionedTransition:    &persistencespb.VersionedTransition{TransitionCount: 4},
+			LastUpdateVersionedTransition: &persistencespb.VersionedTransition{TransitionCount: 4},
+		},
+	}
+	root, err := NewTree(s.registry, persistenceNodes, s.timeSource, s.nodeBackend, s.nodePathEncoder)
+	s.NoError(err)
+
+	// This decoded value should be reset after applying the mutation
+	root.children["child"].value = "some-randome-decoded-value"
+
+	// Prepare mutation: update "child" node, delete "child/grandchild", and add "newchild".
+	updatedChild := &persistencespb.ChasmNode{
+		InitialVersionedTransition:    &persistencespb.VersionedTransition{TransitionCount: 20},
+		LastUpdateVersionedTransition: &persistencespb.VersionedTransition{TransitionCount: 20},
+	}
+	newChild := &persistencespb.ChasmNode{
+		InitialVersionedTransition:    &persistencespb.VersionedTransition{TransitionCount: 100},
+		LastUpdateVersionedTransition: &persistencespb.VersionedTransition{TransitionCount: 100},
+	}
+	mutation := NodesMutation{
+		UpdatedNodes: map[string]*persistencespb.ChasmNode{
+			"child":    updatedChild,
+			"newchild": newChild,
+		},
+		DeletedNodes: map[string]struct{}{
+			"child/grandchild":           {}, // this should remove the entire "grandchild" subtree
+			"child/non-exist-grandchild": {},
+		},
+	}
+	err = root.ApplyMutation(mutation)
+	s.NoError(err)
+
+	// Validate the "child" node got updated.
+	childNode, ok := root.children["child"]
+	s.True(ok)
+	s.Equal(updatedChild, childNode.persistence)
+	s.Nil(childNode.value) // value should be reset after mutation
+
+	// Validate the "newchild" node is added.
+	newChildNode, ok := root.children["newchild"]
+	s.True(ok)
+	s.Equal(newChild, newChildNode.persistence)
+
+	// Validate the "grandchild" node is deleted.
+	s.Empty(childNode.children)
+
+	// Validate that nodeBase.mutation reflects the applied mutation.
+	// Only updates on existing nodes are recorded; new nodes are inserted without a mutation record.
+	expectedMutation := NodesMutation{
+		UpdatedNodes: map[string]*persistencespb.ChasmNode{
+			"child":    updatedChild,
+			"newchild": newChild,
+		},
+		DeletedNodes: map[string]struct{}{
+			"child/grandchild":                 {},
+			"child/grandchild/grandgrandchild": {},
+		},
+	}
+	s.Equal(expectedMutation, root.mutation)
+}
+
+func (s *nodeSuite) TestApplySnapshot() {
```

Comment:
This test won't pass until https://github.com/temporalio/temporal/pull/7477 is landed.